vm/devices/storage/scsidisk/src/lib.rs

```diff
@@ -867,6 +885,134 @@ impl SimpleScsiDisk {
+        // transfer length with atomic boundary
+        if atomic_boundary > 0
+            && len as u32
+                > self
+                    .scsi_parameters
+                    .maximum_atomic_transfer_length_with_atomic_boundary
+                    .unwrap_or(0)
+        {
+            tracelimit::error_ratelimited!(
+                atomic_boundary,
+                maximum_atomic_transfer_length_with_atomic_boundary = self
+                    .scsi_parameters
+                    .maximum_atomic_transfer_length_with_atomic_boundary
+                    .unwrap_or(0),
+                "atomic boundary > 0, transfer length > maximum atomic transfer length with atomic boundary"
+            );
+            return Err(ScsiError::IllegalRequest(AdditionalSenseCode::INVALID_CDB));
+        }
+
+        // If atomic boundary non-zero, make sure less than or equal to max atomic boundary size
+        if atomic_boundary > 0
+            && atomic_boundary as u32
+                > self
+                    .scsi_parameters
+                    .maximum_atomic_boundary_size
+                    .unwrap_or(0)
+        {
+            tracelimit::error_ratelimited!(
+                atomic_boundary,
+                maximum_atomic_boundary_size = self
+                    .scsi_parameters
+                    .maximum_atomic_boundary_size
+                    .unwrap_or(0),
+                "atomic boundary > maximum atomic boundary size"
+            );
+            return Err(ScsiError::IllegalRequest(AdditionalSenseCode::INVALID_CDB));
+        }
+
+        // If atomic transfer length granularity non-zero, make sure that either transfer length OR
+        // the non-zero atomic boundary field is a multiple of this granularity
+        if self
+            .scsi_parameters
+            .atomic_transfer_length_granularity
+            .unwrap_or(0)
+            > 0
+        {
+            if len as u32
+                % self
+                    .scsi_parameters
+                    .atomic_transfer_length_granularity
+                    .unwrap()
+                != 0
+                && (atomic_boundary > 0
+                    && atomic_boundary as u32
+                        % self
+                            .scsi_parameters
+                            .atomic_transfer_length_granularity
+                            .unwrap()
+                        != 0)
+            {
+                tracelimit::error_ratelimited!(
+                    len,
+                    atomic_boundary,
+                    atomic_transfer_length_granularity = self
+                        .scsi_parameters
+                        .atomic_transfer_length_granularity
+                        .unwrap_or(0),
+                    "transfer length OR non-zero atomic boundary not multiple of atomic transfer length granularity"
+                );
+                return Err(ScsiError::IllegalRequest(AdditionalSenseCode::INVALID_CDB));
```

Comment:
I recognize that you are still mulling over the validation strategy here, but this type of thing is useful to include as a unit test (basic packet parsing & protocol validations).